src/fetch/payouts.py

```diff
@@ -0,0 +1,348 @@
+        #  However, there is this issue with the fact that payment can be negative
+        #  (which didn't exist before)
+        # We do not handle overdraft scenario here!
+        assert not self.is_overdraft()
+        if self.total_outgoing_eth() < self.exec_cost:
+            # Total outgoing doesn't even cover execution costs,
+            # so reimburse as much as possible.
+            return (
+                [
+                    Transfer(
+                        token=None,
+                        recipient=self.solver,
+                        amount_wei=int(self.total_outgoing_eth()),
+                    )
+                ]
+                # Handling another Degenerate Case
+                if self.total_outgoing_eth() > 0
+                else []
+            )
+        # Since total_outgoing_eth >= self.exec_cost so there will be some transfer.
+        # TODO - THIS IS A MESS! WAY TOO COMPLICATED TO GET RIGHT IN A SINGLE PR.
+        #  need to mathematically prove that at least one transfer
+        #  will always be returned from this (or find a counter example!)
+        eth_amount = int(self.exec_cost + self.slippage_eth)
+        cow_amount = int(self.total_cow_reward())
+
+        if cow_amount < 0:
+            # This happens when reward_cow < 0 (i.e. payment - cost < 0)
+            # In particular -secondary_reward_cow
+            # So we will need to deduct this from eth_amount
+            assert self.total_eth_reward() < 0
+            return [
+                Transfer(
+                    token=None,
+                    recipient=self.solver,
+                    # eth_amount + self.total_eth_reward()
+                    # = exec_cost + slippage_eth + payment_eth - exec_cost + secondary_reward_eth
+                    # = slippage_eth + payment_eth + secondary_reward_eth
+                    # >= exec_cost >= 0
+                    amount_wei=eth_amount + self.total_eth_reward(),
+                )
+            ]
+
+        # We should be able to prove that there must always be one transfer resulting from this
+        result = []
+        try:
+            result.append(
+                Transfer(
+                    token=None,
+                    recipient=self.solver,
+                    amount_wei=eth_amount,
+                )
+            )
+        except AssertionError:
+            logging.warning(
+                f"Invalid COW Transfer {self.solver} with amount={cow_amount}"
+            )
+        try:
+            result.append(
+                Transfer(
+                    token=Token(COW_TOKEN_ADDRESS),
+                    recipient=self.reward_target,
+                    amount_wei=cow_amount,
+                )
+            )
+        except AssertionError:
+            logging.warning(
+                f"Invalid COW Transfer {self.solver} with amount={cow_amount}"
+            )
+
+        return result
```

Comment:
I discussed the following logic with @harisang:
With the two variables `reimbursement_eth = exec_cost + slippage_eth` and `reward_eth = payment_eth - exec_cost + secondary_reward_eth`:
1) `if reimbursement_eth + reward_eth < 0`: overdraft for total payment in ETH
2) `elif reimbursement_eth > 0 and reward_eth < 0`: reimburse total payment in ETH; it can be interpreted as deducting the owed rewards from gas reimbursement (incl. slippage)
3) `elif reimbursement_eth < 0 and reward_eth > 0`: reimburse total payment in COW; it can be interpreted as deducting the owed reimbursement (incl. slippage) from rewards
4) `else`: reimburse `reimbursement_eth` in ETH and `rewards_cow` in cow

One thing thing that is missing for making my change work is that `reimbursement_eth` is only available in ETH, even though there is a case where its value in COW would be needed.

```suggestion
        reimbursement_eth = int(self.exec_cost + self.slippage_eth)
        # TODO: the following value is not available in COW at the moment
        reimbursement_cow = int(0)
        reward_eth = int(self.total_eth_reward())
        reward_cow = int(self.total_cow_reward())

        if reimbursement_eth > 0 and reward_cow < 0:
            # If the total payment is positive but the total rewards are negative,
            # pay the total payment in ETH. The total payment corresponds to reimbursement,
            # reduced by the negative reward.
            return [
                Transfer(
                    token=None,
                    recipient=self.solver,
                    amount_wei=reimbursement_eth + reward_eth,
                )
            ]
        elif reimbursement_eth < 0 and reward_cow > 0:
            # If the total payment is positive but the total reimbursement is negative,
            # pay the total payment in COW. The total payment corresponds to a payment of rewards,
            # reduced by the negative reimbursement.
            return (
                [
                    Transfer(
                        token=None,
                        recipient=self.solver,
                        amount_wei=reimbursement_cow + reward_cow,
                    )
                ]
            )

        # TODO: filter out zero amounts
        result = []
        try:
            result.append(
                Transfer(
                    token=None,
                    recipient=self.solver,
                    amount_wei=reimbursement_eth,
                )
            )
        except AssertionError:
            logging.warning(
                f"Invalid ETH Transfer {self.solver} with amount={reimbursement_eth}"
            )
        try:
            result.append(
                Transfer(
                    token=Token(COW_TOKEN_ADDRESS),
                    recipient=self.reward_target,
                    amount_wei=cow_amount,
                )
            )
        except AssertionError:
            logging.warning(
                f"Invalid COW Transfer {self.solver} with amount={cow_amount}"
            )

        return result
```

Reply:
Thanks @fhenneke - I can add `reimbursement_cow` to the original data-frame construction. Will merge this first as you suggest and then fix that.